### DISCUSSION TITLE
How to make some routes visible only for users with granted permission (superuser)

### DISCUSSION THREAD

**Discussion by NightSkySK** [fastapi/fastapi#10264] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import timedelta

from apis.utils import OAuth2PasswordBearerWithCookie
from core.config import settings
from core.hashing import Hasher
from core.security import create_access_token
from db.models.users import User
from db.repository.login import get_user
from db.session import get_async_session
from fastapi import APIRouter, Depends, HTTPException, Response, status

# from fastapi.security import OAuth2PasswordBearer
from fastapi.security import OAuth2PasswordRequestForm
from jose import JWTError, jwt
from schemas.tokens import Token
from schemas.users import ShowUser
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter()


async def authenticate_user(
    username: str,
    password: str,
    session: AsyncSession = Depends(get_async_session),
):
    user = await get_user(username=username, session=session)
    # print(user)
    if not user:
        return False
    if not Hasher.verify_password(password, user.hashed_password):
        return False
    return user


@router.post("/token", response_model=Token)
async def login_for_access_token(
    response: Response,
    form_data: OAuth2PasswordRequestForm = Depends(),
    session: AsyncSession = Depends(get_async_session),
):  # added response as a function parameter
    user = await authenticate_user(form_data.username, form_data.password, session)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
        )
    print("user is ", user.email)
    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.email}, expires_delta=access_token_expires
    )
    print("access token is ", access_token)
    response.set_cookie(
        key="access_token", value=f"Bearer {access_token}", httponly=True
    )  # set HttpOnly cookie in response
    return {"access_token": access_token, "token_type": "bearer"}


oauth2_scheme = OAuth2PasswordBearerWithCookie(tokenUrl="/login/token")


async def get_current_user_from_token(
    token: str = Depends(oauth2_scheme),
    session: AsyncSession = Depends(get_async_session),
):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
    )
    try:
        payload = jwt.decode(
            token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM]
        )
        username: str = payload.get("sub")
        print("username/email extracted is ", username)
        if username is None:
            raise credentials_exception
    except JWTError:
        raise credentials_exception
    user = await get_user(username=username, session=session)
    if user is None:
        raise credentials_exception
    print("user is ", user.email)
    return user


@router.get("/me", response_model=ShowUser)
async def current_user(current_user: User = Depends(get_current_user_from_token)):
    return current_user
```

### Description

In another file I have my routing which should be accessible only for "superuser" 

```
superuser_router = APIRouter(include_in_schema=False)


@superuser_router.post("/company/create")
async def create_company(
    company: CompanyCreate,
    session: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(get_current_user_from_token),
):
    if current_user.superuser:
        company = await create_new_company(company=company, session=session)
        return company
    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED, detail="You are not permitted!!!"
    )
```

Also, the limitation of execution this API is only fro superusers and it works perfectly fine. But I can't find any way to hide OpenAPI / Swagger UI behind to be accessible only for login users or to include or exclude some API from schema based on user privileges.

### Operating System

Windows

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

Python 3.9.2

**Comment by chrisK824** [comment]
Hi @NightSkySK , I've dug a bit into it, following the code of FastAPI core for openapi spec creation. I've found that `include_in_schema` is a an APIRoute class attribute, so it is tied with each route during declaration. That said, you can only define it once defining the route.

So I think the only way to control that dynamically  would be to generate your routes with FastAPI's `add_route` method which would allow to set that flag to True/False based on condition.

That said, still what you need is not possible, as your condition is only known during runtime, meaning you would need to create the route during the runtime. I doubt this is possible but I didn't even bother digging into it as this would be absurd.

So, I think what you are asking is not possible. However, OpenAPI and FastAPI come together with all the security schemas, meaning that if your permission dependencies work correctly a given endpoint will be presented, but a user will always be denied if doesn't have the correct permissions. I don't really see an issue about that, not security related anyway. If this is just for isolating documentation for different roles, you could think something else like mounting sub applications each with their own documentations etc.

**Reply by NightSkySK** [reply]
@chrisK824 thank you for the explanation. I now understand the limitations, and why I couldn't find the solution to my question by googling it. I have one more question in this situation, can I hide OpenAPI docs page to have it available only for authenticated users (it doesn't matter which level of privileges)?

**Reply by chrisK824** [reply, marked as answer]
Hi there again.

I did a little search as it was an interesting topic.

First things first. It seems that messing around with API routes instances during runtime is absolutely possible. Found an answer here https://stackoverflow.com/questions/70783994/reload-routes-in-fastapi-during-runtime?fbclid=IwAR0o8-zPoWNt5n6J1rLE4VYtgKR5QehluH8ADwT1sxHDQgFhJJ5vJpKwq04, so it intrigued me. I thought it will break everything, but it seems it doesn't. 

So, as you can change the route instances, you can change as well the include_in_schema attribute. That said, I was able during runtime to do that for any route I wanted using the stored routes in the app router itself, accessing them like that in a middleware:

```
def get_user_permission_level():
    import random
    return random.randint(1, 3)

openapi_paths_permissions = {
    1: ["/open", "/user", "/superuser"],
    2: ["/user", "/superuser"],
    3: ["/superuser"]
}


@api.middleware("http")
async def add_process_time_header(request: Request, call_next):
    permission_level = get_user_permission_level()
    routes_permitted = openapi_paths_permissions[permission_level]
    
    for route in api.router.routes:
        if route.path not in routes_permitted:
            include = False
            logger.info(f"Excluding from schema: {route.path}")
        else:
            include = True
            logger.info(f"Including in schema: {route.path}")
        route.include_in_schema = include
    api.openapi_schema = None
    api.openapi()
    response = await call_next(request)
    return response
```

After finishing with the tweaks for the routes, I'm setting the openapi_schema to None so that the app sees it is missing when calling the class method openapi().

That trick makes the openapi schema update and include only the routes you wish in it. So, aaaalmost what you needed, but no,

So the good news end here, with this discovery for any potential future usage.

Explain:
Even though you are able to mess around with the routes and the schema during the runtime, that means that you are messing around with the main application attributes and hence schema. So instead of every user to see different things, the result would be that any user will "enforce" its permissions on the schema every time they access the API. 

Now that may work for the most part as the job is done in a middleware so basically in any single API request. If not breaking it would most certainly load the shit on your server.

I tried to isolate the process to only be done when accessing the route `/docs` where the openapi is served at, but the middleware didn't like it, as the `await call_next(request)` was breaking due to incompatibility. Didn't have the time or patience to search more on that but from what I saw in the logs it has to do with internal routing and redirects that happen to serve the HTML swagger, hence not just a simple request propagate in the middleware. If you manage to solve that problem then it may be a solution. The breaking code would be the same but the whole route loop tweaking would be done only with a condition when `route.url.path == `/docs`.

Other than that, I tried your second option request. I wasn't able to override the `/docs` endpoint to include dependency injection, however you can do the same thing in a simple middleware like the above. Again, I could not isolate it only on the docs endpoint for the same reason described earlier.

Wish you best of luck

**Comment by WilliamStam** [comment]
so after about 3 weeks and 3 frameworks later... 

lets break down the requirements:
- we need to attach what permissions the user needs to have to the route when we setup the route. (dev time)
- need to check the current user and allow / disallow them if the current user doesnt have that permission (route run time)
- we need to hide routes that the user doesnt have permission on (spec gen time)

so first order of business... how do we attach the permissions to the route, on another framework they had this nifty thing called "guards". you basicaly add a callable to the route, router, app and it runs like a middlware (and in fastapi's terms.. a dependency). when i first jumped away from fastapi (it was i kid you not cause i couldnt get endpoints hidden based on permissions). but i digress.. those guards get attached to the route handler. and you can loop through the guards... and youc an check if the guard is a certain type... and if you defined your guard as in 
```
class Authoprization():
   def __init__(self,permissions=[]):
      self.permissions = permissions
   def __call__(self):
      # boring bit
```
then you can theoreticaly do something like:

```
for row in guards:
   if isinstance(row,Authorization):
      permissions = row.permissions
```

but fastapi doesnt have guards... but it DOES also attach the dependencies to the route handler..

so define a route like

```
@router.get("/authenticated",dependencies=[Authorization()])
async def testing_authenticated() -> str:
    return "authenticated"

@router.get("/authorized", dependencies=[Authorization(permission)])
async def testing_authorized() -> str:
    return "authorized"
```

and we create a route say

```
@router.get('/openapi.json', include_in_schema=False)
async def openapi(request: Request, user: CurrentUser):
    print("----------------------")
    print(user)
    print("----------------------")
    for route in request.app.routes:
        route_parts = list()
        route_parts.append(str(route))
        if hasattr(route,"dependencies") and type(getattr(route,"dependencies")) == list:
            for row in route.dependencies:
                if hasattr(row,"dependency"):
                    if hasattr(row.dependency,"scopes"):
                        scopes = row.dependency.scopes
                    
                        route_parts.append(f"===")
                        route_parts.append(f"{scopes}")
                        
                        if user.id is None:
                            route_parts.append(f"User must be authed")
                            route.include_in_schema = False
                        if not user.has_permissions(scopes):
                            route_parts.append(f"User must have scopes {scopes}")
                            route.include_in_schema = False
        
        route_parts.append(f" route.include_in_schema: {route.include_in_schema}")
        print(" | ".join(route_parts))

    request.app.openapi_schema = None
    openapi = request.app.openapi()
    
    return JSONResponse(openapi)
```

we end up with:

```
APIRoute(path='/profile', name='get_profile', |  route.include_in_schema: True
APIRoute(path='/login', name='login',  |  route.include_in_schema: True
APIRoute(path='/play/authenticated', name='testing_authenticated',  | === | [] |  route.include_in_schema: True
APIRoute(path='/play/authorized', name='testing_authorized', methods=['GET']) | === | [playground] |  route.include_in_schema: True
APIRoute(path='/play/open', name='testing_open', methods=['GET']) |  route.include_in_schema: True
APIRoute(path='/play/woof/open', name='testing_open2', methods=['GET']) | === | [playground] |  route.include_in_schema: True
APIRoute(path='/play/woof/authorized', name='testing_authorized2', methods=['GET']) | === | [playground] | === | [playground.sub] |  route.include_in_schema: True
APIRoute(path='/favicon.ico', name='favicon', methods=['GET']) |  route.include_in_schema: False

```

this is still very WIP at the moment. but it seems to work. since the routes are setup when you call the openapi.json file it shouldnt have too much of a system hit. 

only things to take care of (and im considering using copy here) is that setting the include_in_schema changes would apply to other users as well type thing since its app wide

**Reply by WilliamStam** [reply]
the above solution here https://github.com/tiangolo/fastapi/discussions/10692
